File: ZhouHaiWatchFace/WatchControlLibrary/CommonHelper.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Controls;
using System.Windows.Media.Imaging;
using System.Windows.Media;
using System.Xml.Serialization;
using WatchControlLibrary.Model;
using System.Windows;
using MessageBox = System.Windows.MessageBox;
using Size = System.Windows.Size;
using System.Windows.Forms;
using Brushes = System.Windows.Media.Brushes;
using Point = System.Windows.Point;
using System.Xml;
using System.Configuration;

namespace WatchControlLibrary
{
    public class CommonHelper
    {
        public static string AppPath => Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "ZHWatch");
        public static string AddBtnBackgroundPath => Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "data", "AddBtnBackground.png");

        public static string CurrentPath(string watchName) => Path.Combine(RootPath(watchName), "resources");

        public static string RootPath(string watchName) => Path.Combine(AppPath, watchName);

        public static string GetJsonPath(string watchName) => Path.Combine(CurrentPath(watchName), "watchFace.json");

        public static string GetXmlPath(string watchName) => Path.Combine(CurrentPath(watchName), "watchFace.xml");

        public static string Preview(string watchName) => Path.Combine(CurrentPath(watchName), "_preview");

        public static string Widget(string watchName) => Path.Combine(CurrentPath(watchName), "_widget");

        public static string Ezip(string watchName) => Path.Combine(CurrentPath(watchName), "ezip");

        public static string AbsolutePath(string relative) => Path.Combine(AppDomain.CurrentDomain.BaseDirectory, relative.TrimStart('.', '\\', '/'));

        public static string OutPutPath(string watchName) => Path.Combine(RootPath(watchName), "output");

        public static string OutPutDescriptFile(string watchName) => Path.Combine(OutPutPath(watchName), "description.xml");
        public static string OutPutDescriptJson(string watchName) => Path.Combine(CurrentPath(watchName), "description.json");

        public static string OutPutPathPreview(string watchName) => Path.Combine(RootPath(watchName), "output", "preview");

        public static string OutPutPathPreviewCoverImage(string watchName) => Path.Combine(OutPutPathPreview(watchName), "cover.jpg");
        public static string OutPutPathPreviewIconImage(string watchName) => Path.Combine(OutPutPathPreview(watchName), "icon_small.jpg");

        public static string WatchBin(string watchName) => Path.Combine(RootPath(watchName), "WatchBin");

        public static string EzipBin(string watchName) => Path.Combine(RootPath(watchName), "EzipBin");



        public static string Resource => "resource.bin";

        public static readonly string? Version = ConfigurationManager.AppSettings["Version"];

        public static string[] Languages => new string[]
        {
            "zh_CN",
            "en_US",
            "zh_TW",
            "ja_JP",
            "es_ES",
            "fr_FR",
            "de_DE",
            "ru_RU",
            "pt_BR",
            "pt_PT",
            "it_IT",
            "ko_KR",
            "tr_TR",
            "nl_NL",
            "th_TH",
            "sv_SE",
            "da_DK",
            "vi_VN",
            "nb_NO",
            "pl_PL",
            "fi_FI",
            "in_ID",
            "el_GR",
            "ro_RO",
            "cs_CZ",
            "uk_UA",
            "sk_SK",
            "hu_HU",
            "ar_EG",
            "iw_IL",
            "zh_HK"
        };

        public static string OutPath(string path, string watchName) => path.TrimStart('.')
            .Replace(@"\\", @"\")
            .Replace(@$"\ZHWatch\{watchName}\resources\", "")
            .Replace(@"\", "/");

        public static JsonSerializerSettings Settings => new JsonSerializerSettings
        {
            TypeNameHandling = TypeNameHandling.Auto,
            TypeNameAssemblyFormatHandling = TypeNameAssemblyFormatHandling.Simple,
        };

      

        public static void GenerateDesc(string? folderName)
        {
            var descriptionInfo = CommonHelper.OutPutDescriptJson(folderName);
            if (File.Exists(descriptionInfo))
            {
                var info = JsonConvert.DeserializeObject<WatchInfo>(File.ReadAllText(descriptionInfo), CommonHelper.Settings);
                var watch = new WatchInfoExtra();
                watch.zh.title = info.HnTheme.extra_zh_title;
                watch.zh.briefInfo = info.HnTheme.extra_zh_briefInfo;
                watch.en.title = info.HnTheme.extra_en_title;
                watch.en.briefInfo = info.HnTheme.extra_en_briefInfo;

                var js = JsonConvert.SerializeObject(watch);
                info.HnTheme.version = "qxzhzn.2.0.0";
                info.HnTheme.screen = "QXHD03";
                info.HnTheme.extra = js;
                var fs = CommonHelper.OutPutDescriptFile(folderName);
                XmlSerializer fsserializer = new XmlSerializer(typeof(HnTheme));
                var xmlSettings = new XmlWriterSettings
                {
                    Indent = true,
                    OmitXmlDeclaration = true // 去掉 XML 声明
                };

                using (StreamWriter writer = new StreamWriter(fs))
                using (XmlWriter xmlWriter = XmlWriter.Create(writer, xmlSettings))
                {
                    var namespaces = new XmlSerializerNamespaces();
                    namespaces.Add(string.Empty, string.Empty); // 添加一个空的命名空间
                    fsserializer.Serialize(xmlWriter, info.HnTheme, namespaces);
                }
            }
        }

        public static void GenerateXml(string? folderName)
        {
            var watchFach = JsonConvert.DeserializeObject<WatchFace>(File.ReadAllText(GetJsonPath(folderName)), Settings);
            var outXml = watchFach.GetWatchfaceOut();
            //outXml.WatchInfo = info;
            XmlSerializer serializer = new XmlSerializer(typeof(WatchfaceOut));
            var xmlPath = @$"{CommonHelper.CurrentPath(folderName)}\manifest.xml";
            using (StreamWriter writer = new StreamWriter(xmlPath))
            {
                serializer.Serialize(writer, outXml);
            }
        }

        public static void GenerateImage(string? folderName)
        {

            var watchFace = JsonConvert.DeserializeObject<WatchFace>(File.ReadAllText(GetJsonPath(folderName)), Settings);
            //Directory.Delete(CommonHelper.Preview(watchFace.FolderName), true);
            //Directory.CreateDirectory(CommonHelper.Preview(watchFace.FolderName));
            foreach (var style in watchFace.WatchStyles)
            {
                SetDefaultNum(style.DragBindBases);
                MyCanvas myCanvas = new MyCanvas();
                myCanvas.Width = watchFace.Width;
                myCanvas.Height = watchFace.Height;
                if (watchFace.IsAlbum ?? false)
                {
                    // style.DragBindBases.Insert(0,watchFace.AlbumBackground);
                    myCanvas.Background = watchFace.AlbumBackground!.Background;
                }
                else
                {
                    myCanvas.Background = Brushes.Black;
                }

                if (myCanvas.Width == myCanvas.Height)
                {
                    myCanvas.Clip = new EllipseGeometry
                    {
                        Center = new Point(myCanvas.Width / 2, myCanvas.Height / 2),
                        RadiusX = myCanvas.Width / 2,
                        RadiusY = myCanvas.Height / 2,
                    };
                }
                else
                {
                    myCanvas.Clip = new RectangleGeometry
                    {
                        Rect = new Rect(0, 0, myCanvas.Width, myCanvas.Height),
                        RadiusX = watchFace.Corner,
                        RadiusY = watchFace.Corner,
                    };
                }


                myCanvas.ItemsSource = style.DragBindBases;


                myCanvas.UpdateLayout();

                ExportToPng($@"{Preview(watchFace.FolderName)}\", style.StyleName, myCanvas, watchFace.ThumbnailWidth, watchFace.ThumbnailHeight);



            }
        }




        public static void SetDefaultNum(IEnumerable<DragBindBase> bindBases)
        {
           
            foreach (var bindBase in bindBases)
            {
                var setting = DataHelper.GetWatchSettingByItemName(bindBase.ItemName);

                if (bindBase is DragBindNums bindNums)
                {
                    bindNums.CurrentNum = (int)(setting?.Default ?? 0);
                }
                else if (bindBase is DragBindProgress bindProgress)
                {
                    bindProgress.TargetValue = (int)(setting?.TargetValue ?? 0);

                    bindProgress.CurrentNum = (int)(setting?.Default ?? 0);
                }
                else if (bindBase is DragBindDouble bindDouble)
                {
                    bindDouble.CurrentNum = setting?.Default ?? 0;
                }
                else if (bindBase is DragBindKeyValue bindKeyValue)
                {
                    bindKeyValue.CurrentNum = (int)(setting?.Default ?? 0);
                }
                else if (bindBase is DragBindNormalDateTime dragBindNormalDateTime)
                {
                    dragBindNormalDateTime.SetDateTime = MonitorItem.DefaultTime;

                }
                else if (bindBase is DragBindMonthDay dragMonthDay)
                {
                    dragMonthDay.SetDateTime = MonitorItem.DefaultTime;

                }
                else if (bindBase is DragBindWeek dragBindWeek)
                {
                    dragBindWeek.SetDateTime = MonitorItem.DefaultTime;

                }
                else if (bindBase is DragBindAMPM dragBindAMPM)
                {
                    dragBindAMPM.SetDateTime = MonitorItem.DefaultTime;

                }
                else if (bindBase is DragBindPoint dragPoint)
                {
                    
                    dragPoint.Value = WatchStyle.GetTimeAngle(dragPoint.ValueIndex, MonitorItem.DefaultTime);
                }
                else if (bindBase is DragBindSwitch dragSwitch)
                {
                    dragSwitch.IsOpen = dragSwitch.ItemName == "温度单位" ? true : false;
                }
                else if (bindBase is DragBindSingleDigit singleDigit)
                {
                    singleDigit.SetDateTime = MonitorItem.DefaultTime;
                }
                else if (bindBase is DragBindWidget widget)
                {
                    SetDefaultNum(widget.SubItems);
                }
                else if(bindBase is DragBindSlot slot) 
                {
                    SetDefaultNum(slot.SubItems);
                }
            }

        }

        public static void ExportToPng(string path, string pngName, Canvas surface, int thumbnailWidth = 0, int thumbnailHetght = 0)
        {

            Transform transform = surface.LayoutTransform;
            surface.LayoutTransform = null;


            RenderTargetBitmap renderBitmap = GetCanvasBitmap(surface, (int)surface.Width, (int)surface.Height);  // 效果图
            using (FileStream outStream = new FileStream(path + pngName + ".png", FileMode.OpenOrCreate))
            {
                PngBitmapEncoder encoder = new PngBitmapEncoder();
                encoder.Frames.Add(BitmapFrame.Create(renderBitmap));
                encoder.Save(outStream);
            }

            if (thumbnailWidth == 0 || thumbnailHetght == 0)
                return;
            surface.LayoutTransform = new ScaleTransform
            {
                CenterX = 0.5,
                CenterY = 0.5,
                ScaleX = thumbnailWidth / surface.Width,
                ScaleY = thumbnailHetght / surface.Height,
            };
            renderBitmap = GetCanvasBitmap(surface, thumbnailWidth, thumbnailHetght);  //缩略图
            using (FileStream outStream = new FileStream(path + $"thumbnail_{pngName}" + ".png", FileMode.OpenOrCreate))
            {
                PngBitmapEncoder encoder = new PngBitmapEncoder();
                encoder.Frames.Add(BitmapFrame.Create(renderBitmap));
                encoder.Save(outStream);
            }

            surface.LayoutTransform = transform;
        }


        static RenderTargetBitmap GetCanvasBitmap(Canvas canvas, int w, int h)
        {
            Size size = new Size(w, h);
            canvas.Measure(size);
            canvas.Arrange(new Rect(size));

            RenderTargetBitmap renderBitmap =
            new RenderTargetBitmap(
            (int)size.Width,
            (int)size.Height,
            96d,
            96d,
            PixelFormats.Pbgra32);
            renderBitmap.Render(canvas);
            return renderBitmap;
        }
    }
}

File: WFT-MY-JL/src/models/CommonHelper.ts
/**
 * Common helper utilities equivalent to C# CommonHelper
 * Provides path operations, serialization settings, and utility functions
 */
export class CommonHelper {
  // Path configurations
  public static readonly appPath = "/watch-faces/"; // Web equivalent
  public static readonly addBtnBackgroundPath = "/assets/AddBtnBackground.png";

  // Language codes
  public static readonly languages = [
    "zh_CN", "en_US", "zh_TW", "ja_JP", "es_ES", "fr_FR", "de_DE", "ru_RU",
    "pt_BR", "pt_PT", "it_IT", "ko_KR", "tr_TR", "nl_NL", "th_TH", "sv_SE",
    "da_DK", "vi_VN", "nb_NO", "pl_PL", "fi_FI", "in_ID", "el_GR", "ro_RO",
    "cs_CZ", "uk_UA", "sk_SK", "hu_HU", "ar_EG", "iw_IL", "zh_HK"
  ];

  // Version info
  public static readonly version = "2.0.0";

  /**
   * Get current path for watch resources
   */
  public static currentPath(watchName: string): string {
    return `${this.rootPath(watchName)}/resources`;
  }

  /**
   * Get root path for watch
   */
  public static rootPath(watchName: string): string {
    return `${this.appPath}${watchName}`;
  }

  /**
   * Get JSON path for watch face
   */
  public static getJsonPath(watchName: string): string {
    return `${this.currentPath(watchName)}/watchFace.json`;
  }

  /**
   * Get XML path for watch face
   */
  public static getXmlPath(watchName: string): string {
    return `${this.currentPath(watchName)}/watchFace.xml`;
  }

  /**
   * Get preview path
   */
  public static preview(watchName: string): string {
    return `${this.currentPath(watchName)}/_preview`;
  }

  /**
   * Get widget path
   */
  public static widget(watchName: string): string {
    return `${this.currentPath(watchName)}/_widget`;
  }

  /**
   * Get ezip path
   */
  public static ezip(watchName: string): string {
    return `${this.currentPath(watchName)}/ezip`;
  }

  /**
   * Get absolute path
   */
  public static absolutePath(relative: string): string {
    return relative.startsWith('.') ? relative.substring(1) : relative;
  }

  /**
   * Get output path
   */
  public static outputPath(watchName: string): string {
    return `${this.rootPath(watchName)}/output`;
  }

  /**
   * Get output description file path
   */
  public static outputDescriptFile(watchName: string): string {
    return `${this.outputPath(watchName)}/description.xml`;
  }

  /**
   * Get output description JSON path
   */
  public static outputDescriptJson(watchName: string): string {
    return `${this.currentPath(watchName)}/description.json`;
  }

  /**
   * Get output preview path
   */
  public static outputPathPreview(watchName: string): string {
    return `${this.rootPath(watchName)}/output/preview`;
  }

  /**
   * Get output preview cover image path
   */
  public static outputPathPreviewCoverImage(watchName: string): string {
    return `${this.outputPathPreview(watchName)}/cover.jpg`;
  }

  /**
   * Get output preview icon image path
   */
  public static outputPathPreviewIconImage(watchName: string): string {
    return `${this.outputPathPreview(watchName)}/icon_small.jpg`;
  }

  /**
   * Get watch bin path
   */
  public static watchBin(watchName: string): string {
    return `${this.rootPath(watchName)}/WatchBin`;
  }

  /**
   * Get ezip bin path
   */
  public static ezipBin(watchName: string): string {
    return `${this.rootPath(watchName)}/EzipBin`;
  }

  /**
   * Resource file name
   */
  public static readonly resource = "resource.bin";

  /**
   * Convert path format for output
   */
  public static outPath(path: string, watchName: string): string {
    return path
      .replace(/^\./, '') // Remove leading dot
      .replace(/\\\\/g, '/') // Replace double backslashes with forward slash
      .replace(/\\/g, '/') // Replace backslashes with forward slashes
      .replace(new RegExp(`/watch-faces/${watchName}/resources/`, 'g'), '') // Remove base path
      .replace(/\/+/g, '/'); // Replace multiple slashes with single slash
  }

  /**
   * JSON serialization settings equivalent to C# settings
   */
  public static readonly jsonSettings = {
    // TypeScript/JavaScript equivalent of C# TypeNameHandling settings
    preserveTypes: true,
    includeTypeInfo: true
  };

  /**
   * Generate description file (web equivalent)
   */
  public static async generateDesc(folderName: string): Promise<void> {
    try {
      const descriptionInfo = this.outputDescriptJson(folderName);
      // In web environment, this would involve fetching the JSON file
      // and converting to appropriate format for download
      console.log(`Generating description for: ${descriptionInfo}`);
    } catch (error) {
      console.error('Error generating description:', error);
    }
  }

  /**
   * Generate XML file (web equivalent)
   */
  public static async generateXml(folderName: string): Promise<void> {
    try {
      const jsonPath = this.getJsonPath(folderName);
      // In web environment, this would involve fetching and converting JSON to XML
      console.log(`Generating XML for: ${jsonPath}`);
    } catch (error) {
      console.error('Error generating XML:', error);
    }
  }

  /**
   * Validate file exists (web equivalent)
   */
  public static async fileExists(path: string): Promise<boolean> {
    try {
      const response = await fetch(path, { method: 'HEAD' });
      return response.ok;
    } catch {
      return false;
    }
  }

  /**
   * Read text file (web equivalent)
   */
  public static async readTextFile(path: string): Promise<string> {
    const response = await fetch(path);
    if (!response.ok) {
      throw new Error(`Failed to read file: ${path}`);
    }
    return await response.text();
  }

  /**
   * Download file (web equivalent of file operations)
   */
  public static downloadFile(content: string, filename: string, mimeType: string = 'application/json'): void {
    const blob = new Blob([content], { type: mimeType });
    const url = URL.createObjectURL(blob);
    const link = document.createElement('a');
    link.href = url;
    link.download = filename;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    URL.revokeObjectURL(url);
  }

  /**
   * Parse JSON with error handling
   */
  public static parseJson<T>(jsonString: string): T | null {
    try {
      return JSON.parse(jsonString) as T;
    } catch (error) {
      console.error('Error parsing JSON:', error);
      return null;
    }
  }

  /**
   * Stringify object with proper formatting
   */
  public static stringifyJson(obj: any, indent: number = 2): string {
    return JSON.stringify(obj, null, indent);
  }

  /**
   * Get file extension
   */
  public static getFileExtension(filename: string): string {
    return filename.split('.').pop()?.toLowerCase() || '';
  }

  /**
   * Check if file is image
   */
  public static isImageFile(filename: string): boolean {
    const imageExtensions = ['jpg', 'jpeg', 'png', 'gif', 'bmp', 'webp', 'svg'];
    return imageExtensions.includes(this.getFileExtension(filename));
  }

  /**
   * Generate unique ID
   */
  public static generateId(): string {
    return Math.random().toString(36).substring(2) + Date.now().toString(36);
  }

  /**
   * Deep clone object
   */
  public static deepClone<T>(obj: T): T {
    return JSON.parse(JSON.stringify(obj));
  }

  /**
   * Debounce function
   */
  public static debounce<T extends (...args: any[]) => any>(
    func: T,
    wait: number
  ): (...args: Parameters<T>) => void {
    let timeout: any = null;
    return (...args: Parameters<T>) => {
      if (timeout) clearTimeout(timeout);
      timeout = setTimeout(() => func(...args), wait);
    };
  }

  /**
   * Format file size
   */
  public static formatFileSize(bytes: number): string {
    if (bytes === 0) return '0 B';
    const k = 1024;
    const sizes = ['B', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
  }

  /**
   * Set default values for components (equivalent to C# SetDefaultNum)
   */
  public static setDefaultNum(bindBases: any[]): void {
    for (const bindBase of bindBases) {
      // Get setting from DataHelper
      const setting = (window as any).DataHelper?.getWatchSettingByItemName?.(bindBase.itemName || '');

      // Set default values based on component type
      if (bindBase.constructor?.name === 'DragBindNums') {
        bindBase.currentNum = setting?.Default || 0;
      } else if (bindBase.constructor?.name === 'DragBindProgress') {
        bindBase.targetValue = setting?.TargetValue || 0;
        bindBase.currentNum = setting?.Default || 0;
      } else if (bindBase.constructor?.name === 'DragBindDouble') {
        bindBase.currentNum = setting?.Default || 0;
      } else if (bindBase.constructor?.name === 'DragBindKeyValue') {
        bindBase.currentNum = setting?.Default || 0;
      } else if (bindBase.constructor?.name === 'DragBindNormalDateTime') {
        bindBase.setDateTime = new Date(); // Default time
      } else if (bindBase.constructor?.name === 'DragBindMonthDay') {
        bindBase.setDateTime = new Date();
      } else if (bindBase.constructor?.name === 'DragBindWeek') {
        bindBase.setDateTime = new Date();
      } else if (bindBase.constructor?.name === 'DragBindAMPM') {
        bindBase.setDateTime = new Date();
      } else if (bindBase.constructor?.name === 'DragBindPoint') {
        // Calculate angle based on current time
        const now = new Date();
        const valueIndex = bindBase.valueIndex || 0;
        bindBase.value = this.getTimeAngle(valueIndex, now);
      } else if (bindBase.constructor?.name === 'DragBindSwitch') {
        bindBase.isOpen = bindBase.itemName === '温度单位';
      } else if (bindBase.constructor?.name === 'DragBindSingleDigit') {
        bindBase.setDateTime = new Date();
      } else if (bindBase.constructor?.name === 'DragBindWidget') {
        if (bindBase.subItems) {
          this.setDefaultNum(bindBase.subItems);
        }
      } else if (bindBase.constructor?.name === 'DragBindSlot') {
        if (bindBase.subItems) {
          this.setDefaultNum(bindBase.subItems);
        }
      }
    }
  }

  /**
   * Calculate time angle for pointers (equivalent to WatchStyle.GetTimeAngle)
   */
  private static getTimeAngle(valueIndex: number, dateTime: Date): number {
    switch (valueIndex) {
      case 0: // Hour
        return (dateTime.getHours() % 12) * 30 + dateTime.getMinutes() * 0.5;
      case 1: // Minute
        return dateTime.getMinutes() * 6;
      case 2: // Second
        return dateTime.getSeconds() * 6;
      default:
        return 0;
    }
  }

  /**
   * Generate preview images (web equivalent of GenerateImage)
   */
  public static async generateImage(folderName: string): Promise<void> {
    try {
      // This would involve canvas rendering in a web context
      console.log(`Generating preview images for: ${folderName}`);
      // Implementation would create canvas, render components, and save as images
    } catch (error) {
      console.error('Error generating image:', error);
    }
  }

  /**
   * Export canvas to PNG (web equivalent)
   */
  public static exportToPng(canvas: HTMLCanvasElement, filename: string): void {
    const link = document.createElement('a');
    link.download = filename;
    link.href = canvas.toDataURL('image/png');
    link.click();
  }
}